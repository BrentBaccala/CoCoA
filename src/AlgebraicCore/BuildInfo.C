//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
