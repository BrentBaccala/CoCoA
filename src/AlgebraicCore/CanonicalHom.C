//   Copyright (c)  2007,2009  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/CanonicalHom.H"
#include "CoCoA/RingHom.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/PolyRing.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/FractionField.H"

// #include <iostream>  // for debugging only

namespace CoCoA
{

  RingHom CanonicalHom(const ring& domain, const ring& codomain)
  {
    if (domain == codomain) return IdentityHom(domain);

    // Two easy cases:
    if (IsZZ(domain)) return ZZEmbeddingHom(codomain);
    if (IsQQ(domain)) return QQEmbeddingHom(codomain); // NB result is only a partial hom!!

    // Check codomain first, as this makes it possible to exploit certain "shortcuts"
    if (IsFractionField(codomain))
    {
      if (domain == BaseRing(codomain))
        return EmbeddingHom(codomain);
      else
	return EmbeddingHom(codomain)(CanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      if (domain == CoeffRing(codomain))
        return CoeffEmbeddingHom(codomain);
      else
	return CoeffEmbeddingHom(codomain)(CanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      const QuotientRing QR = codomain;
      if (domain == BaseRing(QR))
        return QuotientingHom(QR);
      else
	return QuotientingHom(QR)(CanonicalHom(domain, BaseRing(QR)));
    }

    CoCoA_ERROR(ERR::CanonicalHomFail, "CanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


  RingHom TmpChainCanonicalHom(const ring& domain, const ring& codomain)
  {
    //    std::clog << " TmpChainCanonicalHom ";
    try { return CanonicalHom(domain, codomain); }
    catch (const CoCoA::ErrorInfo& err) {if (err!=ERR::CanonicalHomFail) throw;}
    if (IsFractionField(codomain))
    {
      //      std::clog << " FrF ";
      return EmbeddingHom(codomain) (TmpChainCanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      //      std::clog << " P ";
      return CoeffEmbeddingHom(codomain) (TmpChainCanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      //      std::clog << " QR ";
      const QuotientRing QR = codomain;
      return QuotientingHom(QR) (TmpChainCanonicalHom(domain, BaseRing(QR)));
    }
    CoCoA_ERROR(ERR::CanonicalHomFail, "TmpChainCanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/CanonicalHom.C,v 1.10 2014/07/08 13:14:40 abbott Exp $
// $Log: CanonicalHom.C,v $
// Revision 1.10  2014/07/08 13:14:40  abbott
// Summary: Removed AsQuotientRing; added new defn of BaseRing
// Author: JAA
//
// Revision 1.9  2014/07/08 08:33:18  abbott
// Summary: Removed AsFractionField
// Author: JAA
//
// Revision 1.8  2014/07/07 12:12:17  abbott
// Summary: Removed AsPolyRing
// Author: JAA
//
// Revision 1.7  2012/02/10 10:26:40  bigatti
// -- changed RingZ.H, RingQ.H --> RingZZ.H, RingQQ.H
//
// Revision 1.6  2012/02/08 15:07:08  bigatti
// -- changed: Z,Q -> ZZ,QQ
//
// Revision 1.5  2011/02/18 12:56:08  bigatti
// -- added TmpChainCanonicalHom
//
// Revision 1.4  2009/07/24 14:21:03  abbott
// Cleaned up include directives, and added some fwd decls.
//
// Revision 1.3  2009/07/24 12:27:46  abbott
// Added some comments.
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.2  2007/03/07 11:18:01  bigatti
// -- fixed CanonicalHom goto flag
//
// Revision 1.1  2007/03/05 21:25:02  cocoa
// New CanonicalHom pseudo-ctor.
//
//
