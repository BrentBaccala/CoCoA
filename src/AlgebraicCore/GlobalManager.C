//   Copyright (c)  2007,2010,2011  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/GlobalManager.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/MemPool.H"
#include "CoCoA/DenseUPolyRing.H" // for Hilbert-Poincare' series
#include "CoCoA/utils.H"

#include "CoCoA/error.H"
#include "TmpHilbertDir/TmpPoincareCPP.H" // for Hilbert-Poincare' series

#include "gmp.h"

#include <algorithm>
using std::min;
using std::max;
#include <cstdlib>
using std::malloc;
using std::realloc;
using std::free;
#include <iostream>
// using std::cerr & std::endl for serious warning in GlobalManager dtor
#include <cstring>
using std::memcpy;


// These 3 fns must have C linkage to work with GMP's mem mgr setter.
extern "C"
{
  void* CoCoA_GMP_alloc(size_t sz);
  void* CoCoA_GMP_realloc(void* ptr, size_t oldsz, size_t newsz);
  void CoCoA_GMP_free(void* ptr, size_t sz);
}

void* CoCoA_GMP_alloc(size_t sz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (sz <= CoCoA::GlobalGMPSliceSize())
    return CoCoA::GlobalGMPPoolPtr()->alloc();
  return malloc(sz);
}

void* CoCoA_GMP_realloc(void* ptr, size_t oldsz, size_t newsz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (oldsz <= CoCoA::GlobalGMPSliceSize() &&
      newsz <= CoCoA::GlobalGMPSliceSize())
    return ptr;

  if (oldsz > CoCoA::GlobalGMPSliceSize() &&
      newsz > CoCoA::GlobalGMPSliceSize())
    return realloc(ptr, newsz);

  const size_t n = min(oldsz, newsz);
  void* dest = CoCoA_GMP_alloc(newsz);
  memcpy(dest, ptr, n);
  CoCoA_GMP_free(ptr, oldsz);
  return dest;
}

void CoCoA_GMP_free(void* ptr, size_t sz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (sz <= CoCoA::GlobalGMPSliceSize())
    CoCoA::GlobalGMPPoolPtr()->free(ptr);
  else
    free(ptr);
}



namespace CoCoA
{

  // Pseudo-ctors for RingZZ and RingQ.
  ring MakeUniqueInstanceOfRingZZ(); // Defined in RingZZ.C.
  FractionField MakeUniqueInstanceOfRingQQ(const ring&); // Defined in RingQ.C.

  // Checking fns to be called immediately before calling dtors for RingQ and RingZZ
  bool RingZZStillInUse(const ring& ZZ);  // Defined in RingZZ.C
  bool RingQQStillInUse(const FractionField& Q);  // Defined in RingQ.C


  // The static members of GlobalManager -- effectively global variables.
  const GlobalManager* GlobalManager::ourGlobalDataPtr = 0;
  std::size_t GlobalManager::GMPSliceSize = 0; // size in bytes of slices in the MemPool (compile-time constant)
  MemPool* GlobalManager::GMPPoolPtr = 0;
  std::size_t GlobalManager::ourHPMaxPower = 100;  // for Hilbert-Poincare' series

  const GlobalManager* GlobalManager::ptr(const char* FnName)
  {
    if (GlobalManager::ourGlobalDataPtr == 0)
      CoCoA_ERROR(ERR::NoGlobalMgr, FnName);
    return GlobalManager::ourGlobalDataPtr;
  }


  GlobalManager::GMPMemMgr::GMPMemMgr(GlobalSettings::AllocatorSetting choice, std::size_t SliceSize)
  {
    if (choice == GlobalSettings::SystemAllocator) return;

    myPoolPtr.reset(new MemPool(SliceSize, "Global GMP MemPool")); // must do this first to be exception safe
    GlobalManager::GMPPoolPtr = myPoolPtr.get();
    GlobalManager::GMPSliceSize = GlobalManager::GMPPoolPtr->mySliceSize();
    mp_get_memory_functions(&myPrevAlloc, &myPrevRealloc, &myPrevFree);
    mp_set_memory_functions(&CoCoA_GMP_alloc, &CoCoA_GMP_realloc, &CoCoA_GMP_free);
  }


  GlobalManager::GMPMemMgr::~GMPMemMgr()
  {
    if (myPoolPtr.get() == 0) return;

    mp_set_memory_functions(myPrevAlloc, myPrevRealloc, myPrevFree);
    GlobalManager::GMPSliceSize = 0;
    GlobalManager::GMPPoolPtr = 0;
  }


  // ----------------------------------------------------------------------

  const std::size_t GlobalSettings::ourDefaultSliceSize = 2*sizeof(long);
  const GlobalSettings::ResidueSetting GlobalSettings::ourDefaultResidueSetting = GlobalSettings::SymmResidues;
  const GlobalSettings::AllocatorSetting GlobalSettings::ourDefaultAllocatorSetting = GlobalSettings::SystemAllocator;


  GlobalSettings::GlobalSettings():
      myResidueSettingHasBeenSet(false),
      myAllocatorSettingHasBeenSet(false),
      mySliceSizeHasBeenSet(false),
      myResidueSetting(ourDefaultResidueSetting),
      myAllocatorSetting(ourDefaultAllocatorSetting),
      mySliceSize(ourDefaultSliceSize)
  {}

  GlobalSettings& GlobalSettings::mySetResidueSetting(ResidueSetting r)
  {
    CoCoA_ASSERT(!myResidueSettingHasBeenSet);
    myResidueSettingHasBeenSet = true;
    myResidueSetting = r;
    return *this;
  }

  GlobalSettings& GlobalSettings::mySetAllocatorSetting(AllocatorSetting a)
  {
    CoCoA_ASSERT(!myAllocatorSettingHasBeenSet);
    myAllocatorSettingHasBeenSet = true;
    myAllocatorSetting = a;
    return *this;
  }

  GlobalSettings& GlobalSettings::mySetSliceSize(std::size_t SliceSize)
  {
    CoCoA_ASSERT(!mySliceSizeHasBeenSet);
    mySliceSizeHasBeenSet = true;
    mySliceSize = SliceSize;
    return *this;
  }



  GlobalSettings GlobalSettings::operator()(std::size_t SliceSize) const
  {
    CoCoA_ASSERT(!mySliceSizeHasBeenSet && myAllocatorSetting != SystemAllocator);
    GlobalSettings ans(*this);
    return ans.mySetSliceSize(SliceSize);
  }


  GlobalSettings operator+(const GlobalSettings& arg1, const GlobalSettings& arg2)
  {
    GlobalSettings ans;
    if (arg1.myResidueSettingHasBeenSet && arg2.myResidueSettingHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "residue setting");
    if (arg1.myResidueSettingHasBeenSet) ans.mySetResidueSetting(arg1.myResidueSetting);
    if (arg2.myResidueSettingHasBeenSet) ans.mySetResidueSetting(arg2.myResidueSetting);

    if (arg1.myAllocatorSettingHasBeenSet && arg2.myAllocatorSettingHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "allocator setting");
    if (arg1.myAllocatorSettingHasBeenSet) ans.mySetAllocatorSetting(arg1.myAllocatorSetting);
    if (arg2.myAllocatorSettingHasBeenSet) ans.mySetAllocatorSetting(arg2.myAllocatorSetting);

    if (arg1.mySliceSizeHasBeenSet && arg2.mySliceSizeHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "GMPAllocator slice size");
    if (arg1.mySliceSizeHasBeenSet) ans.mySetSliceSize(arg1.mySliceSize);
    if (arg2.mySliceSizeHasBeenSet) ans.mySetSliceSize(arg2.mySliceSize);

    return ans;
  }


  const GlobalSettings UseSymmResidues(GlobalSettings().mySetResidueSetting(GlobalSettings::SymmResidues));
  const GlobalSettings UseNonNegResidues(GlobalSettings().mySetResidueSetting(GlobalSettings::NonNegResidues));
  const GlobalSettings UseSystemAllocatorForGMP(GlobalSettings().mySetAllocatorSetting(GlobalSettings::SystemAllocator));
  const GlobalSettings UseGMPAllocator(GlobalSettings().mySetAllocatorSetting(GlobalSettings::GMPAllocator));


  // ----------------------------------------------------------------------

  GlobalManager::ZZQQMgr::ZZQQMgr():
      myRingZZ(MakeUniqueInstanceOfRingZZ()),
      myRingQQ(MakeUniqueInstanceOfRingQQ(myRingZZ))
  {}

  GlobalManager::ZZQQMgr::~ZZQQMgr()
  {
    if (RingZZStillInUse(myRingZZ) || RingQQStillInUse(myRingQQ))
      std::cerr << std::endl
                << "*************************" << std::endl
                << "*** IMMINENT DISASTER ***" << std::endl
                << "*************************" << std::endl
                << std::endl
                << ">>>  CoCoA::GlobalManager being destroyed while CoCoA objects still live!  <<<" << std::endl
                << std::endl;
  }


  // ----------------------------------------------------------------------

  GlobalManager::GlobalManager(const GlobalSettings& settings):
      myResidueSetting(settings.myResidueSetting),
      myGMPMemMgr(settings.myAllocatorSetting, settings.mySliceSize),
      myZZQQMgr()
  {
// !!!***NOT THREAD SAFE***!!!  Must make next 3 lines atomic.
    // Complain if a GlobalManager object has already been created
    if (ourGlobalDataPtr != 0)
      CoCoA_ERROR(ERR::GlobalManager2, "GlobalManager ctor");
    ourGlobalDataPtr = this;
  }


  GlobalManager::~GlobalManager()
  {
// !!!***NOT THREAD SAFE***!!!  Must make check on ourGlobalDataPtr exclusive!
    if (ourGlobalDataPtr == 0)
      CoCoA_ERROR(ERR::SERIOUS, "GlobalManager dtor");
    // Delete registered globals in reverse order
    for (long i=len(myDtorList)-1; i >= 0; --i)
      myDtorList[i].myDtor(myDtorList[i].myPtr);
    ourGlobalDataPtr = 0; // "deregister" the global data
  }


//   bool DefaultResiduesAreSymm()
//   {
//     return GlobalManager::ptr("DefaultResiduesAreSymm")->myResiduesAreSymm;
//   }

  GlobalSettings::ResidueSetting DefaultResidueSetting()
  {
    return GlobalManager::ptr("DefaultResidueSetting")->myResidueSetting;
  }

  
  GlobalManager::PtrDtor::PtrDtor(void* ptr, void (*dtor)(void*)): myPtr(ptr), myDtor(dtor) {}

  void RegisterDtorForGlobal(void* ptr, void (*dtor)(void*))
  {
    GlobalManager::ptr("RegisterDtorForGlobal")->myDtorList.push_back(GlobalManager::PtrDtor(ptr,dtor));
  }


  //----------------------------------------------------------------------
  // pre-computed power list for univariate Hilbert-Poincare Series
  void MakeGlobalHPPowerList(const DenseUPolyRing& P)
  {  
    MakeHPPowerList(GlobalManager::ptr("MakeGlobalHPPowerList")->myHPPowerList,
                    P,
                    GlobalManager::ourHPMaxPower);
  }


  int HPPowerListMaxDeg()
  {  
    return GlobalManager::ptr("HPPowerListMaxDeg")->ourHPMaxPower;
  }

  
  ConstRefRingElem HPPowerList(int exp)
  {
    if (exp>HPPowerListMaxDeg())
      CoCoA_ERROR(ERR::ArgTooBig, "HPPowerList");
    return GlobalManager::ptr("HPPowerList")->myHPPowerList[exp];
  }
  
    
  void CopyHPPower(RingElem& res, int exp)
  {
    if (exp<=HPPowerListMaxDeg())
      res = HPPowerList(exp);
    else
    {
      res = HPPowerList(HPPowerListMaxDeg());
      const DenseUPolyRing HSRing = owner(res);
      for (long i=HPPowerListMaxDeg(); i<exp; ++i)
        HSRing->myMulBy1MinusXExp(raw(res), 1);
    }    
  }
  
  RandomSource& GlobalRandomSource()
  {
    return GlobalManager::ptr("GlobalRandomSource")->myRandomSource;
  }

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/GlobalManager.C,v 1.18 2014/07/09 13:01:17 abbott Exp $
// $Log: GlobalManager.C,v $
// Revision 1.18  2014/07/09 13:01:17  abbott
// Summary: Removed AsDenseUPolyRing
// Author: JAA
//
// Revision 1.17  2014/07/01 12:40:53  bigatti
// -- added CopyHPPower and argument check in HPPowerList
//
// Revision 1.16  2014/04/30 16:07:40  abbott
// Summary: Replaced X.size() by len(X)
// Author: JAA
//
// Revision 1.15  2013/06/17 08:54:02  abbott
// Added RegisterDtorForGlobal.
//
// Revision 1.14  2012/10/15 12:35:24  abbott
// Added  std::  prefix.
//
// Revision 1.13  2012/02/08 13:47:16  bigatti
// -- changed Z,Q --> ZZ,QQ
//
// Revision 1.12  2011/05/19 13:54:48  abbott
// Replaced DefaultResiduesAreSymm by DefaultResidueSetting.
//
// Revision 1.11  2011/05/03 10:03:32  abbott
// Added GlobalRandomSource.
// Internally added GlobalManager::ptr to allow neater implementations.
//
// Revision 1.10  2010/11/17 15:52:33  abbott
// Removed out-of-date include of GMPAllocator.H.
//
// Revision 1.9  2010/11/11 17:45:08  abbott
// Moved GMPMemMgr so that it is a nested class inside GlobalManager.
//
// Revision 1.8  2010/10/29 12:06:41  bigatti
// -- added globals for Hilbert-Poincare' series
//
// Revision 1.7  2010/10/27 20:58:45  abbott
// Major reorganization of GlobalManager and GMPAllocator.
//
// Revision 1.6  2010/10/22 14:03:04  abbott
// Major change to GMPAllocator -- it is now set/activated by the GlobalManager.
// This is a Friday afternoon check-in... hope to check in cleaner code in the
// next few days.
//
// Revision 1.5  2010/09/30 14:28:23  abbott
// Replaced auto_ptrs to RingZ and RingQ by direct values; ctor changed accordingly.
//
// Dtor now checks that ref counts in RingZ and RingQ are correct; if not, a rude
// message is printed on cerr (and the program will probably crash after the
// GlobalManager has been destroyed).
//
// Revision 1.4  2009/12/03 17:40:36  abbott
// Added include directives for ZZ.H (as a consequence of removing
// the directive from ring.H).
//
// Revision 1.3  2009/05/14 09:39:29  abbott
// Added possibility to specify "symmetric" or "non-negative" residues
// in quotients of ZZ.  Affects printing of elements in quotients of ZZ
// (also changed printing of elements in general quotient rings).
// Consequent changes in several tests.
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.2  2007/03/05 21:33:13  cocoa
// Improved/cleaned GlobalManager; added doc too.
//
// Revision 1.1  2007/03/03 14:02:11  bigatti
// -- "foundations" renamed into "GlobalManager"
//
// Revision 1.1  2007/03/02 16:46:28  cocoa
// New foundations object which calls ctors and dtors of global objects.
//
