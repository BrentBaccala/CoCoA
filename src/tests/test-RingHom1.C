//   Copyright (c)  2007  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/CanonicalHom.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/PolyRing.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingFp.H"
#include "CoCoA/RingHom.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/ring.H"
//#include "CoCoA/symbol.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for RingHom
// Z->Q, Q->Q[x,y,z,t], Z->Fp, Q->Fpx, Q[x,y,z,t]->Fp[x], Fp[x]->Fp
// CompositeHom
//----------------------------------------------------------------------

// Handy macro for making assertions.
#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

void program()
{
  GlobalManager CoCoAFoundations(UseNonNegResidues);

  // Create the rings we will be using.
  const FractionField QQ = RingQQ(); // used to fail for non-uniqueness of ZZ
  const QuotientRing Fp = NewZZmod(7);

  const PolyRing QQxyzt = NewPolyRing(QQ, 4);
  const PolyRing Fpx = NewPolyRing(Fp, 1);

//   const ring R1 = NewPolyRing(Qxyzt, symbols("alpha"));
//   const ring R2 = NewPolyRing(NewFractionField(R1), symbols("a"));
//   const ring R3 = NewPolyRing(NewFractionField(R2), symbols("b"));

  // Some simple homomorphisms...
//   RingHom Z2QQ = EmbeddingHom(QQ);
//   RingHom QQ2QQxyzt = CoeffEmbeddingHom(QQxyzt);
//   RingHom Z2Fp = QuotientingHom(Fp);
  RingHom ZZ2QQ = CanonicalHom(RingZZ(), QQ);
  RingHom QQ2QQxyzt = CanonicalHom(QQ, QQxyzt);
  RingHom ZZ2Fp = CanonicalHom(RingZZ(), Fp);
  RingHom QQ2Fpx = InducedHom(QQ, ZZEmbeddingHom(Fpx)); // only partial!
//   RingHom QQxyzt2R3 = ChainCanonicalHom(QQxyzt, R3);

//   TEST_ASSERT( QQxyzt2R3(3) == RingElem(R3, 3) );

  // A homomorphism from the polynomial ring QQxyzt to Fpx.
  vector<RingElem> v1;
  v1.push_back(zero(Fpx));       // x |-> 0
  v1.push_back(indet(Fpx,0));    // y |-> x in Fpx
  v1.push_back(RingElem(Fpx,1)); // z |-> 1 in Fpx
  v1.push_back(zero(Fpx));       // t |-> 0
  RingHom QQxyzt2Fpx = PolyRingHom(QQxyzt, Fpx, QQ2Fpx, v1);

  // A homomorphism from Fpx to Fp.
  vector<RingElem> v2;
  v2.push_back(RingElem(Fp,-1)); // x |-> -1
  RingHom Fpx2Fp = PolyRingHom(Fpx, Fp, IdentityHom(Fp), v2);

  // Compose four homomorphisms
  RingHom CompositeHom = Fpx2Fp(QQxyzt2Fpx(QQ2QQxyzt(ZZ2QQ)));
  cout << "Composite homomorphism is " << CompositeHom << endl << endl;

  for (int i=1; i <= 10; ++i)
  {
    RingElem image = CompositeHom(i); // i is a machine int, and is mapped into ZZ automatically.
    cout << "Using composite hom: " << i << " maps to "
                   << image << " in " << owner(image) << endl;
  }

  cout << "Image of  3  under ZZ2QQ      is " << ZZ2QQ(3) << endl;
  cout << "Image of 3/5 under QQ2QQxyzt  is " << QQ2QQxyzt(RingElem(QQ,3)/5) << endl;
  cout << "Image of 3/5 under QQxyzt2Fpx is " << QQxyzt2Fpx(RingElem(QQxyzt,3)/5) << endl;
  cout << "Image of 3/5 under QQ2Fpx     is " << QQ2Fpx(3/RingElem(QQ,5)) << endl; // works fine
  try
  {
    cout << "Image of 1/7 under QQ2Fpx is " << std::flush;
    cout << QQ2Fpx(1/RingElem(QQ,7)) << endl; // generates an error
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    if (err != ERR::BadRingHomArg2) throw; // rethrow if err is not the expected error.
    cout << "***" << endl
         << "Program correctly detected improper use of a partial ring hom." << endl;
  }
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
