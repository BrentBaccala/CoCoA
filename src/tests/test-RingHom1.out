Composite homomorphism is RingHom(ZZ --> RingWithID(2,"FFp(7)"))

Using composite hom: 1 maps to 1 in RingWithID(2,"FFp(7)")
Using composite hom: 2 maps to 2 in RingWithID(2,"FFp(7)")
Using composite hom: 3 maps to 3 in RingWithID(2,"FFp(7)")
Using composite hom: 4 maps to 4 in RingWithID(2,"FFp(7)")
Using composite hom: 5 maps to 5 in RingWithID(2,"FFp(7)")
Using composite hom: 6 maps to 6 in RingWithID(2,"FFp(7)")
Using composite hom: 7 maps to 0 in RingWithID(2,"FFp(7)")
Using composite hom: 8 maps to 1 in RingWithID(2,"FFp(7)")
Using composite hom: 9 maps to 2 in RingWithID(2,"FFp(7)")
Using composite hom: 10 maps to 3 in RingWithID(2,"FFp(7)")
Image of  3  under ZZ2QQ      is 3
Image of 3/5 under QQ2QQxyzt  is 3/5
Image of 3/5 under QQxyzt2Fpx is 2
Image of 3/5 under QQ2Fpx     is 2
Image of 1/7 under QQ2Fpx is ***
Program correctly detected improper use of a partial ring hom.
